independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/ResteasyReactiveRequestContext.java

```diff
@@ -805,25 +810,46 @@ public Object getHeader(String name, boolean single) {
         }
     }
 
-    @Override
     public Object getQueryParameter(String name, boolean single, boolean encoded) {
+        return getQueryParameter(name, single, encoded, null);
+    }
+
+    @Override
+    public Object getQueryParameter(String name, boolean single, boolean encoded, String separator) {
         if (single) {
             String val = serverRequest().getQueryParam(name);
             if (encoded && val != null) {
                 val = Encode.encodeQueryParam(val);
             }
-            return val;
+
+            if (separator != null) {
+                String[] parts = val.split(separator);
+                return new ArrayList<>(Arrays.asList(parts));
```

Comment:
We're in the `single` branch: we shouldn't split, no?

Reply:
If you look at what happens in the QueryParamExtractor before the PR, we split on the separator even if single is true. So the way it currently is in the PR preserves the behaviour for non-bean query params, whereas your suggestion would change the behaviour in that case.

Of course, it's possible that the old behaviour was wronvlg. What do you think?

Reply:
I see. I suspect the old behaviour was wrong. What would happen if you had a separator, a single value `foo=a,b` and a parameter of type `@RestQuery String foo`? Probably a `ClassCastException` due to trying to inject the `List` into a `String`, no?
You could add a test for that.

Reply:
Yes, you're right. But I do wonder, isn't it always a bug when we attach `@Separator` to a single parameter? Maybe we should just throw an exception during endpoint indexing in this situation?

Reply:
I have now adapted the code to throw an exception during the build when a single parameter is annotated with `@Separator`. Let me know what you think about this.